### DISCUSSION TITLE
uWS::Loop standalone usage

### DISCUSSION THREAD

**Discussion by markmaker** [uNetworking/uWebSockets#1705] [Q&A] [answered]
Hi,

I'm trying to use a `uWS::Loop` standalone, i.e., without an `uWS::App`, in one or more separate threads, so I can `loop->defer()` CPU-bound and/or synchronous tasks to it, exactly the same way that task then defers the response back to the central `uWS::App` loop (`res->send()`, `res->end()` or `ws->send()` inside the defer). A nice symmetric system. :smile: 

After some tweaking, this works nicely, performant, and very robustly even when hammered, but I still have two questions:

1. Calling the standalone `loop->run()` immediately falls through, as just waiting for deferred work does not count as "active". Is there a way to make such a standalone loop remain active?
    Note, just because I'm using  `libuv` (`LIBUS_USE_LIBUV`), I was able to fix this by adding my own active libuv handles to the underlying native uv loop, but that would not work with plain `uWS`, so I'm wondering if there is an official way.
1. Once I got the standalone loop running, I noticed there is a 1-second interval timer installed (my `uv_prepare_t` gets called once a second). Can I somehow get rid of it, or stop it, or  at least set a longer interval? Why is it there?

Thanks,
_Mark

**Comment by uNetworkingAB** [comment]
Why don't you use a proper thread pool? It's simpler and faster

**Comment by markmaker** [comment]
Because in that _other_ thread, I want to use asynchronous techniques too, i.e., I need to handle events _other_ than queued tasks (signal handlers, timers, idle handlers, etc.). Therefore, _some_ kind of loop is a must, and I just thought it would be nice and symmetrical to use the same loop tech.

Furthermore, libuv's `uv_async_t` sucks in that it does not queue. Your defer does. Plus you implement `MoveOnlyFunction` (not that I understand it, or its implications fully yet). :sweat_smile: 

It also happens that I want to use the main thread as the worker. Most thread pools can't do that. Yes, it's a legacy single threaded data-base-centric, forked app that must run tasks of a user-session in tight [serialized ACID transactions](https://en.wikipedia.org/wiki/Concurrency_control#Serializability).

Finally, it's one dependency less. :relieved:

**Reply by uNetworkingAB** [reply]
I don't know what the 1 second timer is, we don't have such a timer. We have a 4 second timer but only if you have an App running.

**Reply by markmaker** [reply]
I had a look, I guess it's the `uWS::Loop::dateTimer` instantiated in `uWS::Loop::create`:

https://github.com/uNetworking/uWebSockets/blob/fc2b7be765983fbce8a3839bc3ff16b51c81b44b/src/Loop.h#L76-L87

Creates a [HTTP date](https://www.rfc-editor.org/rfc/rfc9110#field.date) stamp. Is referenced here:

https://github.com/uNetworking/uWebSockets/blob/fc2b7be765983fbce8a3839bc3ff16b51c81b44b/src/HttpResponse.h#L79

Surely saves some processing time in a heavy loaded server, but could perhaps be moved from `uWS::Loop` to `uWS::App`.

**Reply by uNetworkingAB** [reply, marked as answer]
Ok, all apps on the same Loop have the same Date so that's why it is on the Loop. Forgot about that one

**Reply by markmaker** [reply]
Does make sense. Thanks.
